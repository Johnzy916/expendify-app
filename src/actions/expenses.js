import database from '../firebase/firebase.js';

// ADD_EXPENSES ACTION
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// WRITE EXPENSE TO DATABASE / DISPATCH ACTION
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    
    return database.ref(`users/${uid}/expenses`).push(expense)
      .then(ref => {
        dispatch(addExpense({
          id: ref.key, 
          ...expense 
        }));
      }).catch(error => {
        console.log('Couldn\'t add expense > ', error);
      });
  };
};

// REMOVE_EXPENSE ACTION
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// REMOVE EXPENSE FROM THE DATABASE / DISPATCH ACTION
export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      }).catch(error => console.log('couldn\'t remove expense > ', error));
  };
};

// EDIT_EXPENSE ACTION
export const editExpense = (id, edits) => ({
  type: 'EDIT_EXPENSE',
  id,
  edits
});

// EDIT EXPENSE ON THE DATABASE / DISPATCH ACTION
export const startEditExpense = (id, edits) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(edits)
      .then(() => {
        dispatch(editExpense(id, edits));
    }).catch(error => console.log('couldn\'t update expense > ', error));
  };
};

// SET_EXPENSE ACTION
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// FETCH EXPENSES FROM DATABASE / DISPATCH ACTION
export const startSetExpenses = () => {
 return (dispatch, getState) => {
   const uid = getState().auth.uid;
   return database.ref(`users/${uid}/expenses`).once('value')
  .then(snapshot => {
    const expenses = [];
    snapshot.forEach(expense => {
      expenses.push({
        id: expense.key,
        ...expense.val()
      });
    });
    dispatch(setExpenses(expenses));
  }).catch(error => console.log('couldn\'t get expenses > : ', error));
 };
};

