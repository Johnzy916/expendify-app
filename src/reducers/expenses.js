const expensesReducer = (state, action) => {
  if (typeof state === 'undefined') { 
    return [] 
  }
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.edits
          };
        } else { return expense };
      });
    default:
      return state;
  }
}

export default expensesReducer;