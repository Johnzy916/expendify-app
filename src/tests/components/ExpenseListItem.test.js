import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
import '../setupTests.js'; // needed for codesandbox.io
import toJSON from 'enzyme-to-json'; // needed for codesandbox.io

test('should render Expense List Item with expense', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
  expect(toJSON(wrapper)).toMatchSnapshot();
});