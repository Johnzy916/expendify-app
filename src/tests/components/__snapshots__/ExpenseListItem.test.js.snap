// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List Item with expense 1`] = `
<Link
  className="list-item"
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Pickles
    </h3>
    <span
      className="list-item__subtitle"
    >
      January 5th, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $1.23
  </h3>
</Link>
`;
